File: render.py
#!/usr/bin/env python3
import json
from collections import Counter

COUNTRY_NAMES = {
    "AR": "Argentina", "AT": "Austria", "AU": "Australia", "BE": "Belgium",
    "BR": "Brazil", "CA": "Canada", "CH": "Switzerland", "CL": "Chile",
    "CZ": "Czechia", "DE": "Germany", "DK": "Denmark", "ES": "Spain",
    "FI": "Finland", "FR": "France", "HK": "Hong Kong", "IE": "Ireland",
    "IL": "Israel", "IN": "India", "IS": "Iceland", "IT": "Italy",
    "JP": "Japan", "KR": "South Korea", "MX": "Mexico", "MY": "Malaysia",
    "NL": "Netherlands", "NO": "Norway", "NZ": "New Zealand", "PL": "Poland",
    "PT": "Portugal", "RO": "Romania", "SE": "Sweden", "SG": "Singapore",
    "TR": "Turkey", "TW": "Taiwan", "UK": "United Kingdom",
    "US": "United States", "ZA": "South Africa",
}

THEMES = {
    "light": {
        "ink": "#1f2328", "muted": "#59636e",
        "bar": "#6d4aff", "rule": "#d1d9e0",
    },
    "dark": {
        "ink": "#f0f6fc", "muted": "#9198a1",
        "bar": "#8b6dff", "rule": "#3d444d",
    },
}

FONT = "ui-sans-serif, -apple-system, Segoe UI, Helvetica, Arial, sans-serif"
WIDTH = 880
PADDING = 28
GUTTER = 132
VALUE_SPACE = 62
ROW_HEIGHT = 26
BAR_HEIGHT = 13
BAR_TOP = 152
TOP_COUNTRIES = 10


def read_json(path):
    with open(path, encoding="utf-8") as file:
        return json.load(file)


def count_ranges(path):
    with open(path, encoding="utf-8") as file:
        return sum(1 for line in file if line.strip() and not line.startswith("#"))


def exit_ips_by_country(logicals):
    counts = Counter()
    for logical in logicals["LogicalServers"]:
        servers = [s for s in logical.get("Servers", []) if s.get("ExitIP")]
        counts[logical.get("ExitCountry")] += len(servers)
    counts.pop(None, None)
    return counts


def text(x, y, content, fill, size, weight="400", anchor="start"):
    return (
        f'<text x="{x}" y="{y}" fill="{fill}" font-family="{FONT}" '
        f'font-size="{size}" font-weight="{weight}" text-anchor="{anchor}">'
        f"{content}</text>"
    )


def bar_path(x, y, width, height, radius=4):
    if width <= radius:
        return f'<path d="M{x} {y}h{width}v{height}h{-width}z"/>'
    straight = width - radius
    return (
        f'<path d="M{x} {y}h{straight}a{radius} {radius} 0 0 1 {radius} {radius}'
        f"v{height - 2 * radius}a{radius} {radius} 0 0 1 {-radius} {radius}"
        f'h{-straight}z"/>'
    )


def stat_tiles(stats, theme):
    parts = []
    step = (WIDTH - 2 * PADDING) / len(stats)
    for index, (value, label) in enumerate(stats):
        x = PADDING + index * step
        parts.append(text(x, 52, f"{value:,}", theme["ink"], 30, "600"))
        parts.append(text(x, 76, label, theme["muted"], 13))
    return parts


def country_rows(counts, theme):
    top = counts.most_common(TOP_COUNTRIES)
    largest = top[0][1]
    span = WIDTH - PADDING - GUTTER - VALUE_SPACE
    parts = []
    for index, (code, count) in enumerate(top):
        y = BAR_TOP + index * ROW_HEIGHT
        width = max(2, round(span * count / largest))
        name = COUNTRY_NAMES.get(code, code)
        parts.append(text(PADDING, y + 11, name, theme["ink"], 13))
        parts.append(f'<g fill="{theme["bar"]}">{bar_path(GUTTER, y, width, BAR_HEIGHT)}</g>')
        parts.append(
            text(GUTTER + width + 10, y + 11, f"{count:,}", theme["muted"], 12)
        )
    return parts


def render(stats, counts, theme_name):
    theme = THEMES[theme_name]
    height = BAR_TOP + TOP_COUNTRIES * ROW_HEIGHT + 14
    caption = f"Exit IPs by country, top {TOP_COUNTRIES} of {len(counts)}"
    body = [
        *stat_tiles(stats, theme),
        f'<line x1="{PADDING}" y1="100" x2="{WIDTH - PADDING}" y2="100" '
        f'stroke="{theme["rule"]}" stroke-width="1"/>',
        text(PADDING, 130, caption, theme["muted"], 13),
        *country_rows(counts, theme),
    ]
    return (
        f'<svg xmlns="http://www.w3.org/2000/svg" width="{WIDTH}" '
        f'height="{height}" viewBox="0 0 {WIDTH} {height}" '
        f'role="img" aria-label="{caption}">' + "".join(body) + "</svg>\n"
    )


def main():
    logicals = read_json("protonvpn_logicals.json")
    counts = exit_ips_by_country(logicals)
    stats = [
        (len(read_json("protonvpn_ips.json")), "exit IPs"),
        (len(read_json("protonvpn_entry_ips.json")), "entry IPs"),
        (len(read_json("protonvpn_subdomains.json")), "hostnames"),
        (count_ranges("protonvpn_entry_ip_ranges.txt"), "ASN ranges"),
    ]
    for theme_name in THEMES:
        path = f"assets/overview-{theme_name}.svg"
        with open(path, "w", encoding="utf-8") as file:
            file.write(render(stats, counts, theme_name))
        print(f"wrote {path}")


if __name__ == "__main__":
    main()

File: .github/builder/render.go
package main

import (
	"fmt"
	"maps"
	"slices"
	"strings"
)

const (
	font       = "ui-sans-serif, -apple-system, Segoe UI, Helvetica, Arial, sans-serif"
	width      = 880
	padding    = 28
	gutter     = 96
	valueSpace = 62
	rowHeight  = 26
	barHeight  = 13
	barTop     = 152
	topRows    = 10
)

var themes = map[string]map[string]string{
	"light": {"ink": "#1f2328", "muted": "#59636e", "bar": "#6d4aff", "rule": "#d1d9e0"},
	"dark":  {"ink": "#f0f6fc", "muted": "#9198a1", "bar": "#8b6dff", "rule": "#3d444d"},
}

type stat struct {
	value int
	label string
}

func thousands(value int) string {
	text := fmt.Sprint(value)
	for start := len(text) - 3; start > 0; start -= 3 {
		text = text[:start] + "," + text[start:]
	}

	return text
}

func label(x, y int, content, fill string, size int, weight string) string {
	return fmt.Sprintf(
		`<text x="%d" y="%d" fill="%s" font-family="%s" font-size="%d" `+
			`font-weight="%s">%s</text>`,
		x, y, fill, font, size, weight, content)
}

func bar(x, y, length int, fill string) string {
	radius := min(4, length)
	return fmt.Sprintf(
		`<g fill="%s"><path d="M%d %dh%da%d %d 0 0 1 %d %dv%da%d %d 0 0 1 %d %dh%dz"/></g>`,
		fill, x, y, length-radius, radius, radius, radius, radius,
		barHeight-2*radius, radius, radius, -radius, radius, -(length - radius))
}

func exitIPsByCountry(data logicals) map[string]int {
	counts := map[string]int{}
	for _, logical := range data.LogicalServers {
		for _, server := range logical.Servers {
			if server.ExitIP != "" && logical.ExitCountry != "" {
				counts[logical.ExitCountry]++
			}
		}
	}

	return counts
}

func countryRows(counts map[string]int, theme map[string]string) []string {
	codes := slices.SortedFunc(maps.Keys(counts), func(a, b string) int {
		if counts[a] != counts[b] {
			return counts[b] - counts[a]
		}
		return strings.Compare(a, b)
	})
	codes = codes[:min(topRows, len(codes))]

	span := width - padding - gutter - valueSpace
	rows := make([]string, 0, len(codes)*3)
	for index, code := range codes {
		y := barTop + index*rowHeight
		length := max(2, span*counts[code]/counts[codes[0]])
		rows = append(rows,
			label(padding, y+11, code, theme["ink"], 13, "500"),
			bar(gutter, y, length, theme["bar"]),
			label(gutter+length+10, y+11, thousands(counts[code]), theme["muted"], 12, "400"))
	}

	return rows
}

func overview(stats []stat, counts map[string]int, theme map[string]string) string {
	height := barTop + min(topRows, len(counts))*rowHeight + 14
	caption := fmt.Sprintf("Exit IPs by country, top %d of %d",
		min(topRows, len(counts)), len(counts))

	body := []string{}
	step := (width - 2*padding) / len(stats)
	for index, entry := range stats {
		x := padding + index*step
		body = append(body,
			label(x, 52, thousands(entry.value), theme["ink"], 28, "600"),
			label(x, 76, entry.label, theme["muted"], 13, "400"))
	}

	body = append(body, fmt.Sprintf(
		`<line x1="%d" y1="100" x2="%d" y2="100" stroke="%s" stroke-width="1"/>`,
		padding, width-padding, theme["rule"]))
	body = append(body, label(padding, 130, caption, theme["muted"], 13, "400"))
	body = append(body, countryRows(counts, theme)...)

	return fmt.Sprintf(
		`<svg xmlns="http://www.w3.org/2000/svg" width="%d" height="%d" `+
			`viewBox="0 0 %d %d" role="img" aria-label="%s">%s</svg>`+"\n",
		width, height, width, height, caption, strings.Join(body, ""))
}

func renderOverview(data logicals, stats []stat) {
	counts := exitIPsByCountry(data)
	for name, theme := range themes {
		writeFile(fmt.Sprintf(".github/assets/overview-%s.svg", name),
			[]byte(overview(stats, counts, theme)))
	}
}
